File: DoItTest.Tools/Types/Results/DataResult.cs
namespace DoItTest.Tools.Types.Results
{
    public readonly struct DataResult<T>
    {
        public T? Data { get; }
        public IEnumerable<Error> Errors { get; }
        public Boolean IsSuccess => !Errors.Any();

        private DataResult(T? data, IEnumerable<Error> errors)
        {
            Data = data;
            Errors = errors;
        }

        public static DataResult<T> Success(T data)
        {
            return new(data, new Error[0]);
        }

        public static DataResult<T> Failed(IEnumerable<Error> errors)
        {
            return new DataResult<T>(default, errors);
        }

        public static DataResult<T> Failed(String error)
        {
            return new DataResult<T>(default, new[] { new Error(error) });
        }
    }
}

File: DoItTest/ClientApp/src/tools/results/dataResult.ts
import { ResultError, mapToResultErrors } from "./resultError"

export interface DataResult<T> {
    isSuccess: boolean,
    errors: ResultError[],
    data: T
}

export namespace DataResult {
    export function get<T>(result: any, dataConverter: (data: any) => T): DataResult<T> {
        return { isSuccess: result.isSuccess, errors: mapToResultErrors(result.errors), data: dataConverter(result.data) }
    }
}

export const mapToDataResult = (data: any): DataResult<any> => {
    return { isSuccess: data.isSuccess, errors: mapToResultErrors(data.errors), data: data.data }
}